src/MongoDB/WriteResult.c

```diff
@@ -466,50 +465,111 @@ HashTable *php_phongo_writeresult_get_debug_info(zval *object, int *is_temp TSRM
-	ADD_ASSOC_LONG_EX(&retval, "nUpserted", intern->write_result.nUpserted);
 
-	/* Use native arrays for debugging output */
-	state.map.root_type = PHONGO_TYPEMAP_NATIVE_ARRAY;
-	state.map.document_type = PHONGO_TYPEMAP_NATIVE_ARRAY;
+	if (bson_iter_init_find(&iter, intern->reply, "upserted") && BSON_ITER_HOLDS_ARRAY(&iter)) {
+		uint32_t len;
+		const uint8_t *data;
+		php_phongo_bson_state state = PHONGO_BSON_STATE_INITIALIZER;
+
+		/* Use native arrays for debugging output */
+		state.map.root_type = PHONGO_TYPEMAP_NATIVE_ARRAY;
+		state.map.document_type = PHONGO_TYPEMAP_NATIVE_ARRAY;
 
-	phongo_bson_to_zval_ex(bson_get_data(&intern->write_result.upserted), intern->write_result.upserted.len, &state);
+		bson_iter_array(&iter, &len, &data);
+		phongo_bson_to_zval_ex(data, len, &state);
 #if PHP_VERSION_ID >= 70000
-	ADD_ASSOC_ZVAL_EX(&retval, "upsertedIds", &state.zchild);
+		ADD_ASSOC_ZVAL_EX(&retval, "upsertedIds", &state.zchild);
 #else
-	ADD_ASSOC_ZVAL_EX(&retval, "upsertedIds", state.zchild);
+		ADD_ASSOC_ZVAL_EX(&retval, "upsertedIds", state.zchild);
 #endif
+	} else {
+#if PHP_VERSION_ID >= 70000
+		zval upsertedIds;
+		array_init(&upsertedIds);
+		ADD_ASSOC_ZVAL_EX(&retval, "upsertedIds", &upsertedIds);
+#else
+		zval *upsertedIds = NULL;
+		MAKE_STD_ZVAL(upsertedIds);
+		array_init(upsertedIds);
+		ADD_ASSOC_ZVAL_EX(&retval, "upsertedIds", upsertedIds);
+#endif
+	}
 
-	phongo_bson_to_zval_ex(bson_get_data(&intern->write_result.writeErrors), intern->write_result.writeErrors.len, &state);
+	if (bson_iter_init_find(&iter, intern->reply, "writeErrors") && BSON_ITER_HOLDS_ARRAY(&iter)) {
+		uint32_t len;
+		const uint8_t *data;
+		php_phongo_bson_state state = PHONGO_BSON_STATE_INITIALIZER;
+
+		/* Use native arrays for debugging output */
+		state.map.root_type = PHONGO_TYPEMAP_NATIVE_ARRAY;
+		state.map.document_type = PHONGO_TYPEMAP_NATIVE_ARRAY;
+
+		bson_iter_array(&iter, &len, &data);
+		phongo_bson_to_zval_ex(data, len, &state);
+#if PHP_VERSION_ID >= 70000
+		ADD_ASSOC_ZVAL_EX(&retval, "writeErrors", &state.zchild);
+#else
+		ADD_ASSOC_ZVAL_EX(&retval, "writeErrors", state.zchild);
+#endif
+	} else {
 #if PHP_VERSION_ID >= 70000
-	ADD_ASSOC_ZVAL_EX(&retval, "writeErrors", &state.zchild);
+		zval writeErrors;
+		array_init(&writeErrors);
+		ADD_ASSOC_ZVAL_EX(&retval, "writeErrors", &writeErrors);
 #else
-	ADD_ASSOC_ZVAL_EX(&retval, "writeErrors", state.zchild);
+		zval *writeErrors = NULL;
+		MAKE_STD_ZVAL(writeErrors);
+		array_init(writeErrors);
+		ADD_ASSOC_ZVAL_EX(&retval, "writeErrors", writeErrors);
 #endif
+	}
 
-	if (!bson_empty0(&intern->write_result.writeConcernErrors) &&
-			bson_iter_init(&iter, &intern->write_result.writeConcernErrors) &&
-			bson_iter_next(&iter) &&
-			BSON_ITER_HOLDS_DOCUMENT(&iter)) {
+	if (bson_iter_init_find(&iter, intern->reply, "writeConcernErrors") && BSON_ITER_HOLDS_ARRAY(&iter) &&
+	    bson_iter_recurse(&iter, &child) && bson_iter_next(&child) && BSON_ITER_HOLDS_DOCUMENT(&child)) {
```

Comment:
This "recursion" loop looks very similar to the code for getWriteConcern errors. Is it possible to make a private function for this so that code is not duplicated? The same can be said for the writeErrors and upsertedIds above.